Dockerfile

```diff
@@ -1,13 +1,53 @@
-FROM ubuntu:20.10
-
-RUN apt update -y
-RUN apt install -y \
-    libgtk-3-dev \
-    libglib2.0-dev \
-    libgraphene-1.0-dev \
-    git \
-    xvfb \
-    curl \
-    libcairo-gobject2 \
-    libcairo2-dev \
-    wget
+FROM fedora:rawhide
+
+RUN dnf update -y
+RUN dnf install wget git meson cmake gcc gcc-c++ \
+    libXext-devel libXrender-devel -y
```

Comment:
Shouldn't we also install libpng-dev etc so that we can also build cairo's PNG/etc backends on the CI?

Also `harfbuzz`/`pango` don't need freetype/fontconfig?

Reply:
The way this is right now, libpng, freetype and fontconfig are all built and installed as subprojects of cairo

Reply:
It would seem better to take these dependencies that we don't directly care about from the distro packages, and to only build those ourselves that we a) directly bind, or b) where the distro versions are simply too old / non-existing.

Reply:
And now you have to hope that the different meson subprojects for the same library are building the correct and compatible version for all projects. Installing meson subprojects and sharing the results with other projects (libpng is used at least by cairo and gdk-pixbuf, for example) is super fragile and I'd rather have us build those things completely manually than depending on meson subprojects.

Reply:
I guess we could just install whatever is available in rawhide for now and switch to building manually if needed

Reply:
It's also not like libpng, libjpeg, freetype, fontconfig, etc are projects that are moving fast. The versions we need here at least are probably 10 years old or older :)

Reply:
Why rawhide though? The versions of dependencies we need here are ancient and this Fedora release should have new enough versions for the next few years of GNOME releases.

Reply:
I agree that there is no need to use rawhide if we are building everything we need ourselves anyway...

Reply:
Maybe what I'm writing is confusing, so again a bit clearer :)

  - We should build the libraries we directly bind ourselves, i.e. glib, cairo, graphene, gdk-pixbuf, pango, atk, gtk3/4, etc.
  - We should use system dependencies for basic system libraries like libpng, libjpeg, X11 (already done), Wayland, fontconfig/freetype, etc. instead of relying on the meson subprojects to magically match between projects
  - If system dependencies are too old (this shouldn't really be a problem for anything?) then those should ideally be built separately instead of relying on meson subprojects

Reply:
It can be done for libpng and libjpeg, but fontconfig and freetype cause a problem on fedora: https://github.com/gtk-rs/gtk-rs-core/issues/484#issuecomment-1120282621

Reply:
Out of all of these the only one I think is active enough to need as a subproject for gtk4 is wayland-protocols, for example:

- xdg-activation published Oct 15, 2020: [032a0bdc](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/commit/032a0bdc9b293bfa04a691f92a85ab88d4907447)
- Released in wayland-protocols on Apr 30, 2021: [1.21](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/tags/1.21)
- gtk merges support May 3, 2021: [7b735b88](https://gitlab.gnome.org/GNOME/gtk/-/commit/7b735b88b4a1d47cdb65e4c23ef5a8dfa70d3ff6)
- rawhide gets the 1.21 update on May 5, 2021: [wayland-protocols-devel](https://packages.fedoraproject.org/pkgs/wayland-protocols/wayland-protocols-devel/fedora-35.html)
- The release channel only gets 1.23 on November 2, 2021 in [f35](https://docs.fedoraproject.org/en-US/fedora/f35/release-notes/)